API: deprecate setting of .ordered directly (GH9347, GH9190)

jreback opened the pull request:
closes #9347
closes #9190
closes #9148 

so this is option 4).

Default is now `ordered=False` (independently of whether you specify categories or not)
The ordering is still what factorize does (seen order, if no categories are specified).

Add `set_ordered` to change the `ordered` flag, which by default returns a new object.

```
In [1]: cat = pd.Categorical([0,1,2])
In [2]: cat
Out[2]: 
[0, 1, 2]
Categories (3, int64): [0, 1, 2]

In [3]: cat.ordered
Out[3]: False

In [4]: cat.ordered=True
pandas/core/categorical.py:452: FutureWarning: Setting 'ordered' directly is deprecated, use 'set_ordered'

In [5]: cat = cat.set_ordered(True)

In [6]: cat
Out[6]: 
[0, 1, 2]
Categories (3, int64): [0 < 1 < 2]

In [7]: cat.ordered
Out[7]: True

In [8]: cat = pd.Categorical([0,1,2],ordered=True)

In [9]: cat
Out[9]: 
[0, 1, 2]
Categories (3, int64): [0 < 1 < 2]

In [12]: cat.ordered
Out[12]: True
```

Further added the ability for astype to pass on keywords to the constructor

```
In [1]: Series(["a","b","c","a"]).astype('category',ordered=True)
Out[1]: 
0    a
1    b
2    c
3    a
dtype: category
Categories (3, object): [a < b < c]

In [2]: Series(["a","b","c","a"]).astype('category',categories=list('abcdef'),ordered=False)
Out[2]: 
0    a
1    b
2    c
3    a
dtype: category
Categories (6, object): [a, b, c, d, e, f]
```

Furthermore I put a warning as seemingly simple operations will fail because the user had an _implicity_ ordered Categorical.

```
In [1]: df = DataFrame({ 'A' : Series(list('aabc')).astype('category'), 'B' : np.arange(4) })

In [2]: df['A'].order()
TypeError: Categorical not ordered
you can use .set_ordered(True) to change the Categorical to an ordered one

In [3]: df.groupby('A').sum()
ValueError: cannot sort by an unordered Categorical in the grouper
you can set sort=False in the groupby expression or
make the categorical ordered by using .set_ordered(True)
```

jreback committed:
API: deprecate setting of .ordered directly (GH9347, GH9190)
     add set_ordered method for setting ordered
     default for Categorical is now to NOT order unless explicity specified

jreback committed:
whatsnew doc updates for categorical api changes

jreback committed:
add ability to specify keywords to astype for creation defaults

jreback committed:
fix issue with grouping with sort=True on an unordered Categorical
update categorical.rst docs

test unsortable when ordered=True

jreback commented:
One point to note.

supplying a ``categories`` DOES NOT make the categorical ordered. Ordering is a separate and distinct property.

```
In [3]: pd.Categorical(["a","a","b","b"], categories=["a","b","z"])
Out[3]: 
[a, a, b, b]
Categories (3, object): [a, b, z]

In [4]: pd.Categorical(["a","a","b","b"], categories=["a","b","z"], ordered=True)
Out[4]: 
[a, a, b, b]
Categories (3, object): [a < b < z]
```

jreback commented:
cc @JanSchulz 
@jorisvandenbossche 
@shoyer 
cc @jseabold

JanSchulz commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -298,6 +298,82 @@ Backwards incompatible API changes
+
+      In [2]: df['A'].order()
+      TypeError: Categorical not ordered
+      you can use .set_ordered(True) to change the Categorical to an ordered one
```
the wording suggest that `set_ordered(True)` is inplace and it exposes the underlying object name instead of talking about "categorical data". On the other hand, I tried to get a better wording for 5 min, but couldn't find one :-(

jreback replied:
trying to discourage the ``inplace`` option.

JanSchulz replied:
I just saw that `set_categories()` has the same issue... ok, I drop my case here :-)

JanSchulz commented on pandas/core/categorical.py:
```diff
@@ -263,18 +259,12 @@ def __init__(self, values, categories=None, ordered=None, name=None, fastpath=Fa
-                # "forget" this order, the categorical also is sorted/ordered
-                if ordered is None:
-                    ordered = True
+                codes, categories = factorize(values, sort=ordered)
```
This makes it "not option 4": if the constructor does not get a categories array and ordered=False, it will switch to not-lexi-sorted categories.

jreback replied:
hmm, I think sorting is MORE confusing is it not?

JanSchulz replied:
This was the disccusion in #9346 and especially #9347. This would make the categories sorted by order of appearance if `ordered==False` (which it is now per default). Before the categories were almost always lexi-ordered.

jreback replied:
so ``categories not None`` implies ``ordered=True``?

JanSchulz commented on pandas/core/categorical.py:
```diff
@@ -263,18 +259,12 @@ def __init__(self, values, categories=None, ordered=None, name=None, fastpath=Fa
-                    # give us one by specifying categories
-                    raise TypeError("'values' is not ordered, please explicitly specify the "
-                                    "categories order by passing in a categories argument.")
+                # raise, as we don't have a sortable data structure and so the user should
```
same here, just replace the `if ordered:` part with a warning that categories were not lexi-sorted and the user has to specify one, if he wants to have the categories in a specific order (!= the Categorical ordered)

jreback replied:
This wasn't tested anywhere, can you come up with an example which triggers this?

JanSchulz replied:
This is the test which triggered this codepath:
```
    def test_constructor_unsortable(self):

        # it works!
        arr = np.array([1, 2, 3, datetime.now()], dtype='O')
        factor = Categorical.from_array(arr)
        self.assertFalse(factor.ordered)
```

e.g. this should give different results now:
```
        arr = np.array([1, 2, 3, datetime.now()], dtype='O')
        factor = Categorical(arr, ordered=True)
```
before it errored with a type error, now it either succeeds or errors in the factorize step, but it should print the warning/raises a error that the values are not sortable and the user should supply a user-defined order via supplied `categories=[...]`

jreback replied:
ok, right, but is this ONLY if the user doesn't supply categories AND ``ordered=False`` (which is now the default). Before I think it tried first to factorize (e.g. ``sort=True``), which failed then it warned/raised?

jreback replied:
https://github.com/jreback/pandas/commit/0becba3086289fc452f304e7d628645f9f12a7c7

JanSchulz commented on pandas/core/categorical.py:
```diff
@@ -300,12 +290,7 @@ def __init__(self, values, categories=None, ordered=None, name=None, fastpath=Fa
-                ordered = True
-
-        self.ordered = False if ordered is None else ordered
+        self._ordered = ordered
```
This does not guard against `None`, which was valid before.

jreback replied:
yes I agree, ``None`` is not longer valid

JanSchulz commented on pandas/core/categorical.py:
```diff
@@ -360,7 +345,7 @@ def from_codes(cls, codes, categories, ordered=False, name=None):
         categories : index-like
             The categories for the categorical. Items need to be unique.
-        ordered : boolean, optional
+        ordered : boolean, (defualt False)
```
typo: "default"

JanSchulz commented on pandas/core/categorical.py:
```diff
@@ -460,6 +445,37 @@ def _get_levels(self):
+        warn("Setting 'ordered' directly is deprecated, use 'set_ordered'", FutureWarning)
+        self.set_ordered(value, inplace=True)
+
+    def set_ordered(self, value, inplace=False):
```
I still like "as_ordered/unordered" (or as_nominal/ordinal) more, as that implies that a value is returned and not that it is inplace and "set_ordered" is so near "set_order" which would be the "set to this categories".

jreback replied:
I could buy ``as_ordered()`` and ``as_unordered()``

jreback replied:
changed to using ``as_ordered/as_unordered``

JanSchulz replied:
ok, sounds like a good solution.

JanSchulz commented on pandas/core/generic.py:
```diff
@@ -2207,7 +2207,7 @@ def blocks(self):
         return self.as_blocks()
 
-    def astype(self, dtype, copy=True, raise_on_error=True):
+    def astype(self, dtype, copy=True, raise_on_error=True, **kwargs):
```
+1

JanSchulz commented on pandas/core/groupby.py:
```diff
@@ -1923,8 +1923,16 @@ def __init__(self, index, grouper=None, obj=None, name=None, level=None,
+                # must have an ordered categorical
+                if self.sort:
+                    if not self.grouper.ordered:
+                        raise ValueError("cannot sort by an unordered Categorical in the grouper\n"
```
I haven't read the rest of the file but IMO these two should not raise:

* using a string column as the group_by -> If the column is converted to a Categorical first and the default is kept, then the Categorical is not sortable and you would have to explicitly convert each time (`stringvar.astype("category", ordered=True)`)
* even if the groupby var is a Categorical, I'm fine if this takes the a defined categories order and yields the groupby elements by that unless sort is explicitly set to True.

jreback replied:
- on the first ``df['A'].order()``, that is correct. The ``.order()`` raises telling the user they were relying on an implicity ordered Categorical
- ``sort=True`` is the *default* for grouping, so by definition is *should* raise. The user was using an implicitly ordered Categorical in prior versions, that is why I am noting it

JanSchulz replied:
IMO this should not Error:

```
In [1]: df = DataFrame({ 'A' : Series(list('aabc')).astype('category'), 'B' : np.arange(4) })

In [3]: df.groupby('A').sum()
ValueError: cannot sort by an unordered Categorical in the grouper
you can set sort=False in the groupby expression or
make the categorical ordered by using .set_ordered(True)
```

Or does this error if `A` is a object type without a implemented order?

jreback replied:
it errors ONLY because A is a non-ordered ``Cateorical`` and ``sort`` is the default.

JanSchulz replied:
This will make this awkward:

```
df = DataFrame({ 'A' : list('aabc'), 'C' : np.arange(4) })
df["B"]= df["A"].astype('category')
df.groupby('A').sum() # works
df.groupby('B').sum() # raises
```

JanSchulz replied:
Ok, the much more sensible workaround would be `df["B"]= df["A"].astype('category', ordered=True)` and it even does the right thing if "option 4" happens (if not, this would result in differently ordered groups: A by lexi-sort and B by ordee of appearance...)

jreback commented:
@JanSchulz  https://github.com/jreback/pandas/commit/0becba3086289fc452f304e7d628645f9f12a7c7

jreback commented:
The original discussion in #9347, which the current impl.
 
```
In [14]: cat1 = Series(pd.Categorical(["a","c","b"], ordered=False))

In [16]: cat1
Out[16]: 
0    a
1    c
2    b
dtype: category
Categories (3, object): [a, c, b]

In [15]: cat2 = Series(pd.Categorical(["a","c","b"], ordered=True))

In [17]: cat2
Out[17]: 
0    a
1    c
2    b
dtype: category
Categories (3, object): [a < b < c]

In [19]: cat3 = Series(pd.Categorical(["a","c","b"], categories=['b','c','a'], ordered=True))

In [20]: cat3
Out[20]: 
0    a
1    c
2    b
dtype: category
Categories (3, object): [b < c < a]
```

@JanSchulz does this look right?

JanSchulz commented:
No, the first example is what #9347 is all about and which was kind or rejected by all "option 4" votes:

> 4) leave this alone, but make ordered=None -> ordered=False, e.g. if you don't specify an ordering then your ordering is lexographic but you are not considered ordered (this is the same as 1), but we don't default the order

jreback commented:
ok, so I should leave the sorting alone, then just set the ``ordered`` flag. ok.

jreback commented:
ok, if I change back to what was there before (e.g. this should be option 4)

```
In [4]: cat1
Out[4]: 
0    a
1    c
2    b
dtype: category
Categories (3, object): [a, b, c]

In [5]: cat2
Out[5]: 
0    a
1    c
2    b
dtype: category
Categories (3, object): [a < b < c]

In [6]: cat3
Out[6]: 
0    a
1    c
2    b
dtype: category
Categories (3, object): [b < c < a]
```

JanSchulz commented:
yep

JanSchulz commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -298,6 +298,82 @@ Backwards incompatible API changes
+   cat
+   cat.ordered
+
+   # you can set in the construtor
```
typo: "constru_c_tor"

jreback replied:
done

JanSchulz commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -298,6 +298,82 @@ Backwards incompatible API changes
+   cat
+   cat.ordered
+
+For ease of creation of series of ``Categoricals``, we have added the ability to pass keywords when calling ``.astype()``, these are passed directly to the constructor.
```
...of series of categorical data... That's at least the wording in categorical.rst

jreback replied:
done

JanSchulz replied:
These change also makes this part of categroeical.rst obsolete:

```
By passing a :class:`pandas.Categorical` object to a `Series` or assigning it to a `DataFrame`.
This is the only possibility to specify differently ordered categories (or no order at all) at
creation time and the only reason to use :class:`pandas.Categorical` directly:

.. ipython:: python

    raw_cat = Categorical(["a","b","c","a"], categories=["b","c","d"],
                             ordered=False)
    s = Series(raw_cat)
    s
    df = DataFrame({"A":["a","b","c","a"]})
    df["B"] = raw_cat
    df
```

```
You can also specify differently ordered categories or make the resulting data ordered, by passing these arguments to ``astype()``:

.. ipython:: python

    s = Series(["a","b","c","a"])
    s_cat = s.astype("category", categories=["b","c","d"], ordered=False)
    s_cat

.. note::
    It's also possible to pass a :class:`pandas.Categorical` object directly to a `Series` or assigning it to 
    a `DataFrame`. This is discouraged, as :class:`pandas.Categorical` is an implementation detail 
    which can change.
```

jreback commented:
for completeness

```
In [13]: cat1 = pd.Categorical(["a","c","b"], ordered=False)

In [14]: cat1
Out[14]: 
[a, c, b]
Categories (3, object): [a, b, c]

In [15]: cat2 = pd.Categorical(["a","c","b"], categories=['b','c','a'], ordered=False)

In [16]: cat2
Out[16]: 
[a, c, b]
Categories (3, object): [b, c, a]

In [17]: cat3 = pd.Categorical(["a","c","b"], ordered=True)

In [18]: cat3
Out[18]: 
[a, c, b]
Categories (3, object): [a < b < c]

In [19]: cat4 = pd.Categorical(["a","c","b"], categories=['b','c','a'], ordered=True)

In [20]: cat4
Out[20]: 
[a, c, b]
Categories (3, object): [b < c < a]
```

JanSchulz commented:
These change also makes this part of categroeical.rst obsolete:

```
By passing a :class:`pandas.Categorical` object to a `Series` or assigning it to a `DataFrame`.
This is the only possibility to specify differently ordered categories (or no order at all) at
creation time and the only reason to use :class:`pandas.Categorical` directly:

.. ipython:: python

    raw_cat = Categorical(["a","b","c","a"], categories=["b","c","d"],
                             ordered=False)
    s = Series(raw_cat)
    s
    df = DataFrame({"A":["a","b","c","a"]})
    df["B"] = raw_cat
    df
```

```
You can also specify differently ordered categories or make the resulting data ordered, by passing these arguments to ``astype()``:

.. ipython:: python

    s = Series(["a","b","c","a"])
    s_cat = s.astype("category", categories=["b","c","d"], ordered=False)
    s_cat

.. note::
    It's also possible to pass a :class:`pandas.Categorical` object directly to a `Series` or assigning it to 
    a `DataFrame`. This is discouraged, as :class:`pandas.Categorical` is an implementation detail 
    which can change.
```

JanSchulz commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -298,6 +298,82 @@ Backwards incompatible API changes
+
+.. code-block:: python
+
+   In [1]: cat = pd.Categorical([0,1,2])
```
Not sure if this should be `pd.Categorical(...)` or `Series([...]).astype("category")`. At most places `Categorical` is quite hidden away...

JanSchulz commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -298,6 +298,82 @@ Backwards incompatible API changes
+
+.. _whatsnew_0160.api_breaking.categorical:
+
+In prior versions, Categoricals that with had an unspecified ordering (meaning no ``ordered`` keyword was passed) were defaulted to have a lexographic ordering of the values and designated as ``ordered`` Categoricals. Going forward, the ``ordered`` keyword in the ``Categorical`` constructor will default to ``False``, so ordering must now be explicit.
```
This should either remove the mention of the "lexi sorted categories" or state that `categories are still lexi-sorted, but the resulting categorical data is considered "unordered" (cat.ordered == False)`

jreback commented:
@JanSchulz ok, I think I addressed all of the points and fixed up the docs.

I didn't make ``set_ordered`` an API function (though you can certainly still call it). For the reason you mentioned, possibly user confusing, prob best left off of the public (and .cat API).

jreback commented:
ok fixed up

JanSchulz commented:
I had some more comments, not sure if you saw them. Unfortunately I can't find them anymore due to the forced push :-/

jreback commented:
I changed what u mentioned 

prob need a once over after we merge this anyhow

jorisvandenbossche commented:
@jreback @JanSchulz I don't have time today to look at this in more detail, so if possible, can you keep it open a bit longer?

But already one thing: getting errors with a plain groupby does really not look nice in my opinion. Can't we just sort it in the 'order' of the categories? I know it is not 'ordered' in the sense that the first element is regarded as lower as the last etc, but still, you have the 'order' of the categories in the `categories` attribute.

jreback commented:
So here is a possible compromise. automatically ordering basically violates the API guarantee that we shouldn't allow sorting (or sort-like operations) on unordered Categoricals.

But could simply do an implicit ``.as_ordered()`` inside the operation and warn (OrderingWarning) or something, that what they are doing is bad. I think we could do this for all current raises for non-ordering operations, e.g. in groupby, min, max, order, argosrt, searchsorted.

```
In [1]: df = DataFrame({ 'A' : Series(list('aabc')).astype('category'), 'B' : np.arange(4) })

In [2]: df['A'].order()
pandas/core/categorical.py:1012: UserWarning: Categorical is not ordered
sort will be in the order of the categories
you can use .as_ordered() to change the Categorical to an ordered one

  warn("Categorical is not ordered\n"
Out[2]: 
0    a
1    a
2    b
3    c
Name: A, dtype: category
Categories (3, object): [a, b, c]
```

jreback committed:
on attempting to do an ordering operation on an unordered Categorical,
show an OrderingWarning but let the operation succeed

jreback commented:
latest commit changes all of the raises to ``OrderingWarning``. So users can do what they were doing, now will get a warning.

```
In [1]: s = Series([1,2,3,1], dtype="category")

In [2]: s.min()
pandas/core/categorical.py:843: OrderingWarning: Categorical is not ordered
sort will be in the order of the categories
you can use .as_ordered() to change the Categorical to an ordered one

  OrderingWarning)
Out[2]: 1
```

jreback committed:
v0.16.0.txt / release notes updates

JanSchulz commented:
I don't like the "warning instead of raise"-thing... The old thing (raise in groupby) is not so nice, but at least it's not changing a datatype. The way I think about categorical data is like an user defined integer: you have certain valid values (0...INTMAX), maybe an order (0<1, etc). If we had a "unorderable int", it would be bad if min just issues a warning and than proceeds as if it is a normal int.

The only place I find that "not so nice" is the groupby and I think that is more of a problem of the default of the groupby. 

I see several ways to fix that, but all have drawbacks:

* Use this "warn + change" approach as in the before last commit (see above...).
* keep `ordered=True` as the implicit default -> makes the simple groupby case happy, but not so consistent with R and other expectations
* change to `sort=False` in groupby -> API change :-(
* change to `sort=Null` and only make that False (with a warning) if the input is a unordered Categorical
* simple raise an error for the "simple case", which lets the user know that he should always use explicit ordering information in the `astype("category", ordered=False)`

My preference would be `sort=Null` or `raise of unordered`...

shoyer commented:
I need to look into this more deeply, but my inclination would be to switch to `sort=None` in groupby, which could be interpreted as "sort if possible".  In practice, that would mean "sort everything except unordered categoricals." I would certainly *not* change categorical dtypes in groupby operations if possible.... though in the current implementation (without `CategoricalIndex`), we basically do that anyways.

jreback commented:
this commit [here](https://github.com/jreback/pandas/commit/106b7608f713fe18ed84c9a58905defbffce3208) implementes the ``sort=None`` option.

and then we have the groupby back to 'working' (its doing what it was doing before we changed the default), but you force ``sort=True`` then it WILL raise.

``.order()/min/max`` will still raise

```
In [1]: df = DataFrame({ 'A' : Series(list('aabc')).astype('category'), 'B' : np.arange(4) })

In [2]: df.groupby('A').sum()
Out[2]: 
   B
A   
a  1
b  2
c  3

In [3]: df['A'].order()
TypeError: Categorical not ordered
you can use .as_ordered() to change the Categorical to an ordered one

```

jreback commented:
@jorisvandenbossche @shoyer 

If you guys can have a look today would be gr8. Need this for the rc.

@JanSchulz made the list above.

I think ``sort=None`` is a reasonable approach to this.

jorisvandenbossche commented:
I think there is another option that was not included in @JanSchulz his list for possible fixes:

- allow sorting of unordered categoricals. And then we don't have to change anything in the groupby. 
  This maybe sounds strange, but it is also what R does: it allows sorting, but not max/min eg:

  ```
> cat <- as.factor(c('b', 'a', 'c', 'b'))
> cat
[1] b a c b
Levels: a b c
> sort(cat)
[1] a b b c
Levels: a b c
> min(cat)
Error in Summary.factor(c(2L, 1L, 3L, 2L), na.rm = FALSE) : 
  min not meaningful for factors
> cat2 <- as.ordered(c('b', 'a', 'c', 'b'))
> cat2
[1] b a c b
Levels: a < b < c
> sort(cat2)
[1] a b b c
Levels: a < b < c
> min(cat2)
[1] a
Levels: a < b < c
  ```

  If this is good or bad, I can't really say. But the question maybe is what then is the added value of using an ordered categorical above an unorderd.

But to put in the other way around: why not sorting an unordered categorical? It is a convenience to be able to do this (see eg the groupby case), and what harm does it do that this works without having to convert it to an ordered categorical? Just sorting the values in the order as it is in the ``categories``.

@jreback In the case of using `sort=None` in groupby, what is the order of the result when grouping by an unordered categorical? The order of occurrence?

jreback commented:
@jorisvandenbossche ``sort=None`` -> ``sort=True`` for everything but unordered categoricals. 

I like your soln. It basically doesn't care on groupby's which is fine; they will be ordered as they are ordered now (and not custom ordered like ordered categoricals).

jreback commented:
closing in favor of #9622
